tls/s2n_connection_serialize.c

```diff
@@ -0,0 +1,158 @@
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->cipher_suite->iana_value, S2N_TLS_CIPHER_SUITE_LEN));
+
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->client_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->server_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+
+    POSIX_GUARD(s2n_stuffer_write_uint16(&output, conn->max_outgoing_fragment_length));
+
+    if (conn->actual_protocol_version >= S2N_TLS13) {
+        uint8_t secret_size = 0;
+        POSIX_GUARD(s2n_hmac_digest_size(conn->secure->cipher_suite->prf_alg, &secret_size));
+
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.client_app_secret, secret_size));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.server_app_secret, secret_size));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.resumption_master_secret, secret_size));
+    } else {
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls12.master_secret, S2N_TLS_SECRET_LEN));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->handshake_params.client_random, S2N_TLS_RANDOM_DATA_LEN));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->handshake_params.server_random, S2N_TLS_RANDOM_DATA_LEN));
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_connection_deserialize(struct s2n_connection *conn, uint8_t *buffer, uint32_t buffer_length)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->secure);
+
+    struct s2n_blob context_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&context_blob, buffer, buffer_length));
+    struct s2n_stuffer input = { 0 };
+    POSIX_GUARD(s2n_stuffer_init_written(&input, &context_blob));
+
+    uint64_t context_version = 0;
+    POSIX_GUARD(s2n_stuffer_read_uint64(&input, &context_version));
+    /* No other version is supported currently */
+    POSIX_ENSURE(context_version == S2N_SERIALIZED_CONN_V1, S2N_ERR_INVALID_STATE);
+
+    uint8_t protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN] = { 0 };
+    POSIX_GUARD(s2n_stuffer_read_bytes(&input, protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN));
+    conn->actual_protocol_version = (protocol_version[0] * 10) + protocol_version[1];
+    conn->server_protocol_version = conn->actual_protocol_version;
+    conn->client_protocol_version = conn->actual_protocol_version;
+
+    uint8_t cipher_suite[S2N_TLS_CIPHER_SUITE_LEN] = { 0 };
+    POSIX_GUARD(s2n_stuffer_read_bytes(&input, cipher_suite, S2N_TLS_CIPHER_SUITE_LEN));
+    POSIX_GUARD_RESULT(s2n_cipher_suite_from_iana(cipher_suite, S2N_TLS_CIPHER_SUITE_LEN, &conn->secure->cipher_suite));
+
+    /* Sequence numbers get zeroed during the TLS1.3 key expansion so they are read into temp variables */
+    uint8_t client_sequence_number[S2N_TLS_SEQUENCE_NUM_LEN] = { 0 };
+    uint8_t server_sequence_number[S2N_TLS_SEQUENCE_NUM_LEN] = { 0 };
```

Comment:
Instead of that, maybe we could just wipe the connection if something goes wrong with input parsing? Seems like of inefficient to use a bunch of temp variables.

Reply:
Wiping is definitely worse. Then we're modifying parts of the connection we're not even touching. What's the concern with using temp variables?